parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala

```diff
@@ -0,0 +1,163 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.parboiled2
+
+import org.parboiled2.support.hlist.HList
+
+private[parboiled2] trait ParserMacroMethods {
+
+  /** Converts a compile-time only rule definition into the corresponding rule method implementation.
+    */
+  inline def rule[I <: HList, O <: HList](inline r: Rule[I, O]): Rule[I, O] = ${ ParserMacros.ruleImpl('r) }
+
+  /** Converts a compile-time only rule definition into the corresponding rule method implementation
+    * with an explicitly given name.
+    */
+  inline def namedRule[I <: HList, O <: HList](name: String)(inline r: Rule[I, O]): Rule[I, O] = ${
+    ParserMacros.nameRuleImpl('name)('r)
+  }
+
+}
+
+private[parboiled2] trait RuleRunnable {
+
+  /** THIS IS NOT PUBLIC API and might become hidden in future. Use only if you know what you are doing!
+    */
+  implicit class Runnable[L <: HList](rule: RuleN[L]) {
+    def run()(implicit scheme: Parser.DeliveryScheme[L]): scheme.Result = ???
+  }
+}
+
+object ParserMacros {
+  import scala.quoted._
+  import scala.compiletime._
+
+  def ruleImpl[I <: HList: Type, O <: HList: Type](r: Expr[Rule[I, O]])(using Quotes): Expr[Rule[I, O]] =
+    nameRuleImpl(Expr("todo"))(r)
+
+  def nameRuleImpl[I <: HList: Type, O <: HList: Type](
+      name: Expr[String]
+  )(r: Expr[Rule[I, O]])(using Quotes): Expr[Rule[I, O]] = {
+    import quotes.reflect.*
+
+    def opTreeF(rule: Expr[Rule[I, O]]): OpTree = rule match {
+      case '{
+            (${ lhs }: Rule[I, O])
+              .~((${ rhs }: Rule[I, O]))($c, $d)
+          } =>
+        Sequence(Seq(opTreeF(lhs), opTreeF(rhs)))
+      case '{ ($p: Parser).ch($c) } =>
+        CharMatch(p, c)
+      case '{ ($p: Parser).str($s) } =>
+        StringMatch(p, s)
+      case _ => reportError("Invalid rule definition", r)
+    }
+
+    val opTree: OpTree = opTreeF(r)
+
+    val parser = opTree.parser
+    '{
+      def wrapped: Boolean = ${ opTree.render(wrapped = true) }
+      val matched =
+        if ($parser.__inErrorAnalysis) wrapped
+        else ${ opTree.render(wrapped = false) }
+      if (matched) org.parboiled2.Rule.asInstanceOf[Rule[I, O]] else null
+    }
+  }
+
+  private def reportError(error: String, expr: Expr[Any])(using quotes: Quotes): Nothing = {
+    quotes.reflect.report.error(error, expr)
+    throw new Exception(error)
+  }
+
+  sealed trait OpTree {
+    def parser: Expr[Parser]
+    def render(wrapped: Boolean)(using Quotes): Expr[Boolean]
+  }
+
+  case class Sequence(ops: Seq[OpTree]) extends OpTree {
+    val parser = ops.head.parser
+    override def render(wrapped: Boolean)(using Quotes): Expr[Boolean] =
+      ops
+        .map(_.render(wrapped))
+        .reduceLeft((l, r) => '{ val ll = $l; if (ll) $r else false })
+  }
+
+  sealed abstract class TerminalOpTree(parser: Expr[Parser]) extends OpTree {
```

Comment:
```suggestion
  sealed abstract class TerminalOpTree extends OpTree {
```

(the `parser` method is inherited from `OpTree`)

Reply:
👍 
https://github.com/sirthias/parboiled2/pull/274/commits/86a585ae47d1b45b2f213a43692a0d9da9b7a6e2